include/maths/CLocalOutlierFactors.h

```diff
@@ -502,6 +520,7 @@ class MATHS_EXPORT CLocalOutlierFactors {
             this->run(m_K, std::move(points), scores);
         }
 
+        //! \warning This must be thread safe.
```

Comment:
I wonder: for this implementation of `add` and others all over the place in this change, `add` could be _const_, that would make the thread safeness a bit clearer.

(I know you can always construct a case where this is not true or vice versa)

I am not familiar with the code but maybe: is there a chance of making `add` _const_ in the base class and or have a _const_ version of `CMethod`?

Reply:
The reason that this code is thread safe is slightly subtle and this comment was not really helpful on reflection. Basically, different threads only read/write distinct elements in pre-sized vectors. I'm going to comment this better in the code which calls the add functions.